src/main/java/org/quiltmc/loader/api/config/Config.java

```diff
@@ -0,0 +1,423 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.api.config;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Consumer;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.quiltmc.loader.api.config.values.ValueList;
+import org.quiltmc.loader.api.config.values.ValueMap;
+import org.quiltmc.loader.api.config.values.ValueTreeNode;
+import org.quiltmc.loader.impl.config.ConfigImpl;
+
+@ApiStatus.NonExtendable
+public interface Config {
+	String getModId();
+
+	String getId();
+
+	/**
+	 * The path this config will be saved in, relative to Quilt Loader's config directory.
+	 *
+	 * @return a save path
+	 */
+	Path getSavePath();
+
+	/**
+	 * Adds a listener to this {@link Config} that's called whenever any of its values is updated
+	 *
+	 * @param callback an update listener
+	 */
+	void registerCallback(UpdateCallback callback);
+
+	/**
+	 * @return the metadata attached to this value for the specified type
+	 */
+	<M> M metadata(MetadataType<M, ?> type);
+
+	/**
+	 * @return whether or not this value has any metadata of the specified type
+	 */
+	<M> boolean hasMetadata(MetadataType<M, ?> type);
+
+	/**
+	 * Save this config and all its values to disk
+	 */
+	void serialize();
+
+	/**
+	 * Returns all values held by this config file
+	 *
+	 * <p>For all nodes, including section nodes, see {@link #nodes}
+	 *
+	 * @return all values held by this config file
+	 */
+	Iterable<TrackedValue<?>> values();
+
+	/**
+	 * @param key an iterable of key components to access
+	 * @return the value contained by this config class with key components
+	 */
+	TrackedValue<?> getValue(Iterable<String> key);
+
+	/**
+	 * @return all nodes of the value tree represented by this config file, including section nodes
```

Comment:
You reference "value tree" a lot here, but I can't ever find a place that it's actually defined what that means. The javadoc for ValueTreeNode just says "An element in a config tree.", which is *even more confusing*, because now there's two phrases that are equally useless on describing what the "value/config tree" actually is, and what the difference between the two are

Reply:
I've added more docs regarding the behavior of this method that should help clarify things.